cores/esp8266/StreamDev.cpp

```diff
@@ -0,0 +1,333 @@
+
+    This library is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+    Lesser General Public License for more details.
+
+    You should have received a copy of the GNU Lesser General Public
+    License along with this library; if not, write to the Free Software
+    Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+
+    parsing functions based on TextFinder library by Michael Margolis
+*/
+
+
+#include <Arduino.h>
+#include <StreamDev.h>
+
+using esp8266::polledTimeout::oneShotFastMs;
+using esp8266::polledTimeout::periodicFastMs;
+
+size_t Stream::toFull(Print* to,
+                      const ssize_t len,
+                      const int readUntilChar,
+                      const oneShotFastMs::timeType timeoutMs)
+{
+    setWriteError(STREAMTO_SUCCESS);
+
+    if (len == 0)
+    {
+        return 0;    // conveniently avoids timeout for no requested data
+    }
+
+    // There are two timeouts:
+    // - read (network, serial, ...)
+    // - write (network, serial, ...)
+    // However
+    // - getTimeout() is for reading only
+    // - there is no getOutputTimeout() api
+    // So we use getTimeout() for both,
+    // (also when inputTimeoutPossible() is false)
+
+    // "neverExpires (default, impossible)" is translated to default timeout
+    oneShotFastMs timedOut(timeoutMs >= oneShotFastMs::neverExpires ? getTimeout() : timeoutMs);
+
+    // yield about every 5ms (XXX SHOULD BE A SYSTEM-WIDE CONSTANT?)
+    periodicFastMs yieldNow(5);
+
+    size_t written = 0;
+
+    // len==-1 => maxLen=0 <=> until starvation
+    size_t maxLen = std::max((ssize_t)0, len);
+
+    if (peekBufferAPI())
+
+        // peek-buffer API
+
+        while (!maxLen || written < maxLen)
+        {
+            size_t avpk = availableForPeek();
+            if (avpk == 0 && !inputTimeoutPossible())
+            {
+                // no more data to read, ever
+                break;
+            }
+
+            size_t w = to->availableForWrite();
+            if (w == 0 && !outputTimeoutPossible())
+            {
+                // no more data can be written, ever
+                break;
+            }
+
+            w = std::min(w, avpk);
+            if (maxLen)
+            {
+                w = std::min(w, maxLen - written);
+            }
+            if (w)
+            {
+                const char* directbuf = peekBuffer();
+                bool foundChar = false;
+                if (readUntilChar >= 0)
+                {
+                    const char* last = (const char*)memchr(directbuf, readUntilChar, w);
+                    if (last)
+                    {
+                        w = std::min((size_t)(last - directbuf), w);
+                        foundChar = true;
+                    }
+                }
+                if (w && ((w = to->write(directbuf, w))))
```

Comment:
So here we take the raw buffer from the source and pass it to the dest to write directly.  I'm wondering if there are cases where the source could be in IRAM/PMEM and the destination write goes boom?  Maybe I'm overthinking.  For String it's always byte-addressible RAM, same for File and Client....

Reply:
The peek-buffer API is able to tell whether it is usable (data accessible and block-readable).

For example, if you get to re-read the [`in_flash` argument in `StreamPtr` constructors](https://github.com/esp8266/Arduino/pull/6979/files#diff-b7808f496813d60af97f5316fe6bdd213dcc706b86dee15da068f4015c2fa62cR164), peek-buffer API can be prevented when data are not byte-addressable (`in_flash == true`, see `::peekBufferAPI()` in this class). But the `::read()` functions are still optimized with `memcpy_P()`.
This class allows to `StreamPtr(F("I am in flash")).toAll(Serial)`.
In this case, a pointer to the source (peek-buffer API) will not be provided but the fall-back regular copy through buffer  using overloaded `::read(buffer)` will be faster thanks to `memcpy_P()`.

Note that everything is hidden to the user. A derived class knowing the peek-buffer API can take advantage of it when it says it can be used, otherwise the class will work with legacy Stream API (with possibly optimized overloaded function like above).

Reply:
@earlephilhower You may as a native speaker agree with me that `addressible` is incorrect, so please, @d-a-v, substitute `addressable` for it wherever applicable. A minor nuisance but I tripped over it :-)

Reply:
@dok-net just re-read the last line of @earlephilhower's comment above.
Also, [this](https://en.wiktionary.org/wiki/addressible).